core/src/Pos/Core/JsonLog/JsonLogT.hs

```diff
@@ -0,0 +1,181 @@
+Stability:    experimental
+Portability:  GHC
+
+This module provides the monad transformer @'JsonLogT'@
+for adding JSON logging to a monad transformer stack.
+-}
+
+module Pos.Core.JsonLog.JsonLogT
+    ( JsonLogT
+    , runWithoutJsonLogT
+    , runJsonLogT
+    , runJsonLogT'
+    , runWithJsonLogT
+    , runWithJsonLogT'
+    , JsonLogConfig(..)
+    , jsonLogDefault
+    ) where
+
+import           Universum
+
+import           Control.Concurrent.MVar (MVar, withMVar)
+import           Control.Monad.Base (MonadBase)
+import           Control.Monad.Fix (MonadFix)
+import           Control.Monad.IO.Class (MonadIO (..))
+import           Control.Monad.Morph (MFunctor (..))
+import           Control.Monad.Trans.Class (MonadTrans)
+import           Control.Monad.Trans.Control (MonadBaseControl (..))
+import           Control.Monad.Trans.Lift.Local (LiftLocal)
+import           Control.Monad.Trans.Reader (ReaderT (..))
+import           Data.Aeson (ToJSON, encode)
+import           Data.ByteString.Lazy.Char8 (hPutStrLn)
+import           Formatting (sformat, shown, (%))
+import           System.IO (Handle, hFlush)
+import           System.Wlog (CanLog, HasLoggerName (..), WithLogger,
+                     logWarning)
+
+import           Pos.Core.JsonLog.CanJsonLog (CanJsonLog (..))
+import           Pos.Core.JsonLog.Event (JLTimedEvent, timedIO, toEvent)
+import           Pos.Core.Mockable.Channel (ChannelT)
+import           Pos.Core.Mockable.Concurrent (Promise, ThreadId)
+import           Pos.Core.Mockable.Metrics (Counter, Distribution, Gauge)
+import           Pos.Core.Mockable.SharedAtomic (SharedAtomicT)
+import           Pos.Core.Mockable.SharedExclusive (SharedExclusiveT)
+
+data JsonLogConfig
+    = JsonLogDisabled
+    | JsonLogConfig (MVar Handle) (JLTimedEvent -> IO Bool)
+
+-- | Monad transformer @'JsonLogT'@ adds support for JSON logging
+-- to a monad transformer stack.
+newtype JsonLogT m a = JsonLogT (ReaderT JsonLogConfig m a)
+    deriving (Functor, Applicative, Monad, MonadTrans, MonadIO, MFunctor,
+              MonadThrow, MonadCatch, MonadMask, MonadFix, MonadBase b, LiftLocal)
+
+instance MonadBaseControl b m => MonadBaseControl b (JsonLogT m) where
+
+    type StM (JsonLogT m) a = StM m a
+
+    liftBaseWith f = liftBaseWith f
+
+    restoreM = restoreM
+
+instance WithLogger m => CanLog (JsonLogT m) where
+
+instance WithLogger m => HasLoggerName (JsonLogT m) where
+
+    askLoggerName = lift askLoggerName
+
+    modifyLoggerName f = hoist (modifyLoggerName f)
+
+
+
+
+
+type instance Gauge (JsonLogT m) = Gauge m
+type instance Counter (JsonLogT m) = Counter m
+type instance Distribution (JsonLogT m) = Distribution m
+type instance ThreadId (JsonLogT m) = ThreadId m
+type instance Promise (JsonLogT m) = Promise m
+type instance SharedAtomicT (JsonLogT m) = SharedAtomicT m
+type instance SharedExclusiveT (JsonLogT m) = SharedExclusiveT m
+type instance ChannelT (JsonLogT m) = ChannelT m
+
+
+
+
+
+
```

Comment:
Wow, that is a lot of empty lines. Is that just the way it was in the original file?

Reply:
It seems that was indeed the case. Weird!

Reply:
Yep 🤷‍♂️ In the spirit of minimal changes I left it. But several of these modules definitely could use tidying up.